### PR TITLE
fix: add migration for transferprocess_properties default value

### LINKED ISSUE #296: Older TransferProcess objects cause NullPointerException
## Describe the bug

- The latest `TransferProcess` migration adds the `transferprocess_properties` column as nullable and does not set a default value for it, making all the existing tuples have this column as `NULL`.
The `TransferProcess.getProperties` method does not verify this attribute's nullability when it tries to turn it into an unmodifiable map causing a NullPointerException to be thrown.

### To Reproduce

1. Through DMA fetch a `TransferProcess` object having the `transferprocess_properties` column as NULL (e.g. 0.1.6 objects).

### Expected behavior

- The older objects returned with the property as an empty map.

## Context Information

- Used version: 0.3.0

## Possible Implementation

- New migration defining the new coloumn's default value to `{}` and/or verifying nullability in code.

### PR DESCRIPTION
**Pull request by bcronin90**
## WHAT

Add default value for transferprocess-properties to SQL migration.

## WHY

See Issue #296. Changing Transferprocess.getProperties is out of scope for this repository, so the other approach was taken.

Closes #296

### CODE REVIEW

File: edc-extensions/postgresql-migration/src/main/resources/org/eclipse/tractusx/edc/postgresql/migration/transferprocess/V0_0_6__Snapshot_20230109_Update.sql [hunk 1]
```diff
@@ -9,7 +9,10 @@
 --
 --  Contributors:
 --       ZF Friedrichshafen AG - Add column for the dynamically HTTP Receiver callback endpoints
+--       Mercedes-Benz Tech Innovation GmbH - Add default properties
 --
 
 -- add column
 ALTER TABLE edc_transfer_process ADD COLUMN transferprocess_properties TEXT;
+-- set default value
+ALTER TABLE edc_transfer_process ADD CONSTRAINT transferprocess_properties DEFAULT "{}";
```
Comment (florianrusch-zf) [inline edc-extensions/postgresql-migration/src/main/resources/org/eclipse/tractusx/edc/postgresql/migration/transferprocess/V0_0_6__Snapshot_20230109_Update.sql:18]: Missing empty line at EOF
```suggestion
ALTER TABLE edc_transfer_process ADD CONSTRAINT transferprocess_properties DEFAULT "{}";

```

**Comment by florianrusch-zf** [comment]
@bcronin90 could you change the PR title prefix to `fix`? `bug` is not one of the suggestions in our [`pr_etiquette.md`](https://github.com/eclipse-tractusx/tractusx-edc/blob/main/pr_etiquette.md)

And additionally, you changed an existing migration - wouldn't that mean, that if I upgrade my already existing EDC, that this migration would not been executed, because this migration was already been executed in the past? Don't we need a new migration?

**Comment by bcronin90** [comment]
> @bcronin90 could you change the PR title prefix to `fix`? `bug` is not one of the suggestions in our [`pr_etiquette.md`](https://github.com/eclipse-tractusx/tractusx-edc/blob/main/pr_etiquette.md)
> 
> And additionally, you changed an existing migration - wouldn't that mean, that if I upgrade my already existing EDC, that this migration would not been executed, because this migration was already been executed in the past? Don't we need a new migration?

Changed the title. Yes, currently this would require re-running the migration. I can switch it to a new migration, but I have one issue that I'm not sure how to resolve then: Since the changes happen at the same version, how do I make sure the migrations are executed in the correct order, for a user who hasn't run it yet. Since Line 18 needs to be run after line 16.

**Comment by florianrusch-zf** [comment]
Can't we simply create a new `v0.0.7....` migration. This will been executed automatically after the existing ones. Or not? 😅

**Comment by bcronin90** [comment]
> Can't we simply create a new `V0.0.7....` migration file? This will been executed automatically after the existing ones. Or not? sweat_smile

I was unaware our versioning allowed for that. On it!

File: edc-extensions/postgresql-migration/src/main/resources/org/eclipse/tractusx/edc/postgresql/migration/transferprocess/V0_0_7__Default_Value_For_Properties.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+--
+--  This program and the accompanying materials are made available under the
+--  terms of the Apache License, Version 2.0 which is available at
+--  https://www.apache.org/licenses/LICENSE-2.0
+--
+--  SPDX-License-Identifier: Apache-2.0
+--
+--  Contributors:
+--       Mercedes-Benz Tech Innovation GmbH - Add default properties
+--
```
Comment (florianrusch-zf) [inline edc-extensions/postgresql-migration/src/main/resources/org/eclipse/tractusx/edc/postgresql/migration/transferprocess/V0_0_7__Default_Value_For_Properties.sql:12]: Please fix the copyright ;-)

**Review by florianrusch-zf — APPROVED** [review]

**Comment by paullatzelsperger** [comment]
trying this out in a local postgres quickly shows that the migration should actually be: 
```sql
ALTER TABLE edc_transfer_process ALTER COLUMN transferprocess_properties SET DEFAULT '{}';
```

**Comment by florianrusch-zf** [comment]
> did no-one try this out locally?

Nope, I trusted our pipelines & @bcronin90

File: edc-extensions/postgresql-migration/src/main/resources/org/eclipse/tractusx/edc/postgresql/migration/transferprocess/V0_0_7__Default_Value_For_Properties.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+--  https://www.apache.org/licenses/LICENSE-2.0
+--
+--  SPDX-License-Identifier: Apache-2.0
+--
+--  Contributors:
+--       Mercedes-Benz Tech Innovation GmbH - Rewrite to be SQL Init Schema
+--
+
+-- set default value
+ALTER TABLE edc_transfer_process ADD CONSTRAINT transferprocess_properties DEFAULT "{}";
```
Comment (florianrusch-zf) [inline edc-extensions/postgresql-migration/src/main/resources/org/eclipse/tractusx/edc/postgresql/migration/transferprocess/V0_0_7__Default_Value_For_Properties.sql:15]:
```suggestion
ALTER TABLE edc_transfer_process ALTER COLUMN transferprocess_properties SET DEFAULT '{}';
```

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

**Comment by bcronin90** [comment]
I'm in an internal thing today. Will get back to you tomorrow. ---- On Di, 16 Mai 2023 15:08:33 +0200  ***@***.***  wrote ----
trying this out in a local postgres quickly shows that the migration should actually be:
ALTER TABLE edc_transfer_process ALTER COLUMN transferprocess_properties SET DEFAULT '{}';

**Comment by bcronin90** [comment]
Thanks for the suggestion. Should be correct, but I can't look at this today. Will check tomorrow ---- On Di, 16 Mai 2023 15:10:58 +0200  ***@***.***  wrote ----
@florianrusch-zf requested changes on this pull request.



In edc-extensions/postgresql-migration/src/main/resources/org/eclipse/tractusx/edc/postgresql/migration/transferprocess/V0_0_7__Default_Value_For_Properties.sql:
> @@ -0,0 +1,15 @@
+--
+--  Copyright (c) 2022 Mercedes-Benz Tech Innovation GmbH
+--
+--  This program and the accompanying materials are made available under the
+--  terms of the Apache License, Version 2.0 which is available at
+--  https://www.apache.org/licenses/LICENSE-2.0
+--
+--  SPDX-License-Identifier: Apache-2.0
+--
+--  Contributors:
+--       Mercedes-Benz Tech Innovation GmbH - Rewrite to be SQL Init Schema
+--
+
+-- set default value
+ALTER TABLE edc_transfer_process ADD CONSTRAINT transferprocess_properties DEFAULT "{}";

⬇️ Suggested change
-ALTER TABLE edc_transfer_process ADD CONSTRAINT transferprocess_properties DEFAULT "{}";
+ALTER TABLE edc_transfer_process ALTER COLUMN transferprocess_properties SET DEFAULT '{}';

**Comment by paullatzelsperger** [comment]
@florianrusch-zf #359 fixes the markdown errors

**Pull request merged**